packages/runtime/src/next/api-handler/handlers/proxy-preview-mode.ts

```diff
@@ -85,71 +112,53 @@ async function proxyPreviewModeApiRouteHandler(
-    proxyReq.removeHeader('X-Makeswift-Preview-Mode')
-
-    // The following headers are Next.js-specific and are removed to prevent Next.js from
-    // short-circuiting requests, breaking routing.
-    proxyReq.removeHeader('X-Invoke-Path')
-    proxyReq.removeHeader('X-Invoke-Query')
+  const proxyHeaders = mapRequestHeadersToHeaders(req.headers)
+  const proxyUrl = mapRequestToProxyUrl(req)
 
-    const url = new URL(proxyReq.path, 'http://n')
-
-    url.searchParams.delete('x-makeswift-preview-mode')
-
-    proxyReq.path = url.pathname + url.search
-  })
-
-  if (req.query.secret !== apiKey) return res.status(401).send('Unauthorized')
-
-  const host = req.headers.host
-  const originalCookies = req.headers.cookie
-
-  if (host == null) return res.status(400).send('Bad Request')
-
-  const forwardedProtoHeader = req.headers['x-forwarded-proto']
-
-  let forwardedProto: string[] = []
-  if (Array.isArray(forwardedProtoHeader)) {
-    forwardedProto = forwardedProtoHeader
-  } else if (typeof forwardedProtoHeader === 'string') {
-    forwardedProto = forwardedProtoHeader.split(',')
-  }
-
-  const isForwardedProtoHttps = forwardedProto.includes('https')
-
-  const forwardedSSL = req.headers['x-forwarded-ssl']
-  const isForwardedSSL = typeof forwardedSSL === 'string' && forwardedSSL === 'on'
-
-  const proto = isForwardedProtoHttps || isForwardedSSL ? 'https' : 'http'
-  let target = `${proto}://${host}`
-
-  // If the user generates a locally-trusted CA for their SSL cert, it's likely that Node.js won't
-  // trust this CA unless they used the `NODE_EXTRA_CA_CERTS` option
-  // (see https://stackoverflow.com/a/68135600). To provide a better developer experience, instead
-  // of requiring the user to provide the CA to Node.js, we don't enforce SSL during local
-  // development.
-  const secure = process.env['NODE_ENV'] === 'production'
+  proxyUrl.searchParams.delete('x-makeswift-preview-mode')
+  proxyHeaders.delete('x-makeswift-preview-mode')
+  // The following headers are Next.js-specific and are removed to prevent Next.js from
+  // short-circuiting requests, breaking routing.
+  proxyHeaders.delete('x-invoke-path')
+  proxyHeaders.delete('x-invoke-query')
 
   const previewData: MakeswiftPreviewData = {
     makeswift: true,
+    // This will eventually be dynamic
     siteVersion: MakeswiftSiteVersion.Working,
   }
-  const setCookie = res.setPreviewData(previewData).getHeader('Set-Cookie')
-  res.removeHeader('Set-Cookie')
+
+  const setCookie = res.setPreviewData(previewData).getHeader('set-cookie')
+  res.removeHeader('set-cookie')
 
   if (!Array.isArray(setCookie)) return res.status(500).send('Internal Server Error')
 
-  const additionalCookies = parse(setCookie)
-    .map(cookie => serialize(cookie.name, cookie.value, cookie as CookieSerializeOptions))
-    .join(';')
+  setCookie.forEach((cookie) => proxyHeaders.append('cookie', cookie))
+
+  const response = await fetch(proxyUrl, {
+    referrer: req.headers.referer,
```

Comment:
How come it's necessary to specify the referrer here?

Reply:
I guess it doesn't really matter in preview mode. I was thinking it will send `about:client` as the referrer, but we also send it in `headers` so it's actually duplicative.

Reply:
Yeah, I thought the headers would take care of that.